### PR TITLE
Implement Trapped Rain Water using Dynamic Programming Strategy.

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by TruongNhanNguyen** [comment]
Let's take a look on this PR @realstealthninja & @Panquesito7

File: DIRECTORY.md
Comment (mrmagic2020) [inline DIRECTORY.md:1]: Isn't the directory file automatically updated by [this action](https://github.com/TheAlgorithms/C-Plus-Plus/actions/workflows/directory_writer.yml)? @realstealthninja

File: dynamic_programming/trapped_rainwater.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/**
+ * @file
+ * @brief Implementation of the Trapped Rainwater Problem
```
Comment (mrmagic2020) [inline dynamic_programming/trapped_rainwater.cpp:3]: Maybe include a link to a reliable source explaining the algorithm/implementation, as suggested [here](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program).

**Review by mrmagic2020 — COMMENTED** [review]
The code looks pretty solid and is well-documented. Not sure if modification of `DIRECTORY.md` is needed. What do you think @realstealthninja?

File: DIRECTORY.md
Comment (TruongNhanNguyen) [reply DIRECTORY.md:1]: Yeah, I will revert the change on this file. The `DIRECTORY.md` file should be updated by this automated triggered [directory_writer](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/.github/workflows/directory_writer.yml) workflow.

File: dynamic_programming/trapped_rainwater.cpp (hunk 1 above)
Comment (TruongNhanNguyen) [reply dynamic_programming/trapped_rainwater.cpp:3]: Yeah, I will add links to the algorithm explanation to the documentation.

**Review by mrmagic2020 — APPROVED** [review]
👍

**Pull request closed without merging**